File: 0039_Combination Sum/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        combination = []
        self.dfs(candidates, target, [], combination)
        return combination
    
    def dfs(self, nums, target, path, combination):
        if target == 0:
            combination.append(path)
            return
        
        if target < 0:
            return
        
        for i in range(len(nums)):
            self.dfs(nums[i:], target - nums[i], path + [nums[i]], combination)

File: 0039_Combination Sum/solution.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> combination;
        vector<int> path;
        dfs(candidates, target, path, combination);
        return combination;
    }
    
    void dfs(vector<int> nums, int target, vector<int>& path, vector<vector<int>>& combination) {
        if (target == 0) {
            combination.push_back(path);
            return;
        }
        if (target < 0) {
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            target += -nums[i];
            path.push_back(nums[i]);
            dfs(vector<int>(nums.begin() + i, nums.end()), target, path, combination);
            path.pop_back();
            target += nums[i];
        }
    }
};